### DISCUSSION TITLE
input.portaudio stopped working after upgrading liquidsoap in windows

### DISCUSSION THREAD

**Discussion by bASEstr** [savonet/liquidsoap#3641] [User Support]
**Describe the bug**
input.portaudio now throws the following error after updating from 1.4.X version: 

`Source input.portaudio failed while streaming: Failure("Portaudio error in read_stream: Can't read from an output only stream")!
`
**To Reproduce**
.liq file used:

```
environment.set("PA_RECOMMENDED_INPUT_DEVICE","0") #input device id
stream = buffer(input.portaudio())

output.ao(clock_safe=true, fallible=true, options=[("id", "0")], stream)
```

working .liq in liquidsoap 1.4.1:
```
setenv("PA_RECOMMENDED_INPUT_DEVICE","0") #input device id
stream = buffer(input.portaudio())

output.ao(clock_safe=true, fallible=true, options=[("id", "0")], stream)
```

**Expected behavior**
input.portaudio to read audio input from the device configured in the env varialbe PA_RECOMMENDED_INPUT_DEVICE like it does in 1.4.1

**Version details**

- OS: Windows 10 Pro
- Version 22H2

**Install method**
downloaded binary files from https://github.com/savonet/liquidsoap/releases

**Comment by toots** [comment]
Hi! On windows you should be able to use `output.portaudio` directly. You can use the `--list-portaudio-devices` command line option to list all available portaudio devices.

**Reply by bASEstr** [reply]
Hello Toots!
The output is not a concern, the given script in the post was just for demonstration purposes. 
I am concerned by the fact that after updating liquidsoap I am no longer able to input audio from the configured portaudio device.

Any direction towards getting input.portaudio to work again would be highly appreciated.

**Reply by toots** [reply]
Hi my bad for the output. The input should work the same, though: find the relevant device id from this command line and use it as the `device_id` parameter for `input.portaudio`.

**Comment by bASEstr** [comment]
> Hi my bad for the output. The input should work the same, though: find the relevant device id from this command line and use it as the `device_id` parameter for `input.portaudio`.

Hi!
Didn't even realize device_id was added to the input.portaudio method! I can confirm that the following test script does work:
```
stream = buffer(input.portaudio(device_id=0))

output.ao(clock_safe=true, fallible=true, options=[("id", "0")], stream)
```
Is it a concern that without the parameter device_id input.portaudio seems to want to use an "output only device" like speakers instead of using either the PA_RECOMMENDED_INPUT_DEVICE env or the first available input device?
The documentation has the following to say about the parameter:

> device_id (of type int?, which defaults to null): Device ID. Uses default device if null.

And yet in my original post when I didn't use the device_id parameter, it did not seem to use the default input device, but maybe the output device (hence the error)?
The behavior without the parameter is weird nonetheless.

Thanks again for the help.
